ethereum/contracts/BasicInboundChannel.sol

```diff
@@ -2,77 +2,62 @@
 pragma solidity ^0.8.5;
 pragma experimental ABIEncoderV2;
 
-import "./ParachainLightClient.sol";
-import "./BeefyLightClient.sol";
-import "./SimplifiedMMRVerification.sol";
+import "./ParachainClient.sol";
 
 contract BasicInboundChannel {
     uint256 public constant MAX_GAS_PER_MESSAGE = 100000;
     uint256 public constant GAS_BUFFER = 60000;
 
     uint64 public nonce;
 
-    BeefyLightClient public beefyLightClient;
+    ParachainClient public parachainClient;
+
+    struct MessageBundle {
```

Comment:
Shifting to nonce per message bundle instead of nonce per message is a significant change (even though code-wise, looks pretty simple).

I think it makes sense and great that we've got that in now incidentally too, thanks!

Having said that, just want to think about/discuss the implications of the change a little more, as I haven't thought about it in a while - lets discuss it on our next call. Probably all good though.

Any other thoughts/comments from your side on this?

Reply:
In terms of replay protection, our claim is now no longer "Your message will only be played once because each message nonce is increasing" but rather "Your message will only be played once because each bundle nonce is increasing AND we guarantee that your message will be included in EXACTLY one bundle, no more, no less"

Is that correct?

For the latter guarantee, it depends on new code changes to the parachain code. I have reviewed them and they look good, but can you also confirm that the new pallet code now guarantees that "your message will be included in EXACTLY one bundle, no more, no less", in all scenarios, irrespective of what interactions come into the channel pallets?

Reply:
cc @alistair-singh for a second review on the pallet code to ensure the pallets provide this guarantee

Reply:
Will check :+1:

Reply:
Yeah is fine. If the app pallets (ETHApp, etc) submit a message to the outbound channel pallet, then its guaranteed to only exist in the next upcoming commitment. Upon creating a commitment, the channel's message queue buffer is converted into a message bundle, a commitment is inserted into the header, and then the message queue buffer is deleted.

Reply:
thx!